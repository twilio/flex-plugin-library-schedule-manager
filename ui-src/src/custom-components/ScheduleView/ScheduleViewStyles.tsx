import { styled } from '@twilio/flex-ui';

export const ScheduleViewWrapper = styled('div')`
  display: flex;
  height: 100%;
  flex-flow: column;
  flex-grow: 1;
  flex-shrink: 1;

  .editor {
    width: 40vw;
    position: fixed;
    right: 0;
    top: 52px;
    background: white;
    z-index: 2;
    height: 95vh;
  }
`;

export const ScheduleViewHeader = styled('div')`
  display: flex;
  flex-direction: column;
  padding: 1em;

  strong, h2 {
    font-weight: 600;
    color: inherit;
  }
`;

export const RulesContainer = styled('div')`
  width: 100%;
  overflow: auto;
  max-height: 25vh;

  h4 {
    background-color: rgb(244, 244, 246);
    padding: 1em 0 1em 2em;
    font-weight: 600;
  }
`;

export const PublishModalContent = styled('div')`
  display: flex;
  justify-content: center;
  padding-top: 3em;
`;

export const PublishActionContainer = styled('div')`
  margin: 1em;
`;

export const TableContainer = styled('div')`
  overflow: auto;
  height: 80%;
  margin: 0 1em;
  box-sizing: border-box;

  table {
    border: 1px solid #e1e3ea;
  }
  td:nth-of-type(3),
  th:nth-of-type(3) {
    width: 30%;
  }
  td .disabled {
    color: #e1e3ea;
  }
`;