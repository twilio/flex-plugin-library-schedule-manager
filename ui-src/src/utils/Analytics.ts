import FlexTelemetry from '@twilio/flex-ui-telemetry';
import packageJSON from '../../package.json';

export enum Event {
  OPHRS_SCHEDULE_LIST = 'OpHrs Schedule List',
  OPHRS_CREATE_SCHEDULE = 'OpHrs Create Schedule',
  OPHRS_EDIT_SCHEDULE = 'OpHrs Edit Schedule',
  OPHRS_DUPLICATE_SCHEDULE = 'OpHrs Duplicate Schedule',
  OPHRS_SAVE_SCHEDULE = 'OpHrs Save Schedule',
  OPHRS_DELETE_SCHEDULE = 'OpHrs Delete Schedule',
  OPHRS_PUBLISH_SCHEDULE = 'OpHrs Publish Schedules',
  OPHRS_RESET_SCHEDULE = 'OpHrs Reset Schedules',
  OPHRS_RULE_LIST = 'OpHrs Rule List',
  OPHRS_CREATE_RULE = 'OpHrs Create Rule',
  OPHRS_EDIT_RULE = 'OpHrs Edit Rule',
  OPHRS_DUPLICATE_RULE = 'OpHrs Duplicate Rule',
  OPHRS_SAVE_RULE = 'OpHrs Save Rule',
  DELETE_RULE = 'Delete Rule',
}

export const analytics = new FlexTelemetry({
  source: 'flexui',
  role: packageJSON.name,
  plugin: packageJSON.name,
  pluginVersion: packageJSON.version,
  originalPluginName: packageJSON.id,
});
