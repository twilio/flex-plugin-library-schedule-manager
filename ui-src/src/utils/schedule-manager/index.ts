import { Manager, Notifications } from '@twilio/flex-ui';
import { NotificationIds } from '../../flex-hooks/notifications/ScheduleManager';
import ScheduleManagerService from '../serverless/ScheduleManager/ScheduleManagerService';
import { Rule, Schedule, ScheduleManagerConfig } from '../../types/schedule-manager';
import { ErrorManager, FlexPluginErrorType } from '../../ErrorManager';
import { RRule } from 'rrule';
import { DateTime } from 'luxon';

let config = {
  data: {
    rules: [],
    schedules: [],
  },
  version: '',
} as ScheduleManagerConfig;

const delay = async (ms: number): Promise<void> => {
  return await new Promise((resolve) => setTimeout(resolve, ms));
};

export const canShowScheduleManager = (manager: Manager) => {
  const { roles } = manager.user;
  return roles.indexOf('admin') >= 0;
};

export const loadScheduleData = async (): Promise<ScheduleManagerConfig | null> => {
  const listSchedulesResponse = await ScheduleManagerService.list();

  if (listSchedulesResponse) {
    config = listSchedulesResponse;
  }

  return listSchedulesResponse;
};

export const referencedSchedules = (schedules: Schedule[], rule: Rule): string[] => {
  const refSchedules = [] as string[];
  if (schedules) {
    schedules.forEach((schedule) => {
      if (rule && schedule.rules.indexOf(rule.id) >= 0) {
        refSchedules.push(schedule.name);
      }
    });
  }

  return refSchedules;
};
export const updateScheduleData = (newSchedule: Schedule | null, existingSchedule: Schedule | null): Schedule[] => {
  if (existingSchedule === null && newSchedule !== null) {
    // adding schedule
    console.log('config.data.schedules', config.data.schedules);
    config.data.schedules = [...config.data.schedules, newSchedule];
  } else if (existingSchedule !== null && newSchedule === null) {
    // removing existing schedule
    const existingIndex = config.data.schedules.indexOf(existingSchedule);

    if (existingIndex >= 0) {
      config.data.schedules[existingIndex].isDeleted = true;
      // config.data.schedules.splice(existingIndex, 1);
    }
  } else if (existingSchedule !== null && newSchedule !== null) {
    // updating existing schedule
    const existingIndex = config.data.schedules.indexOf(existingSchedule);

    if (existingIndex >= 0) {
      config.data.schedules.splice(existingIndex, 1, newSchedule);
    }
  }

  return config.data.schedules;
};

export const updateRuleData = (newRule: Rule | null, existingRule: Rule | null): Rule[] => {
  if (existingRule === null && newRule !== null) {
    // adding rule
    config.data.rules = [...config.data.rules, newRule];
  } else if (existingRule !== null && newRule === null) {
    // removing existing rule
    const existingIndex = config.data.rules.indexOf(existingRule);

    if (existingIndex >= 0) {
      config.data.rules[existingIndex].isDeleted = true;
      // config.data.rules.splice(existingIndex, 1);
    }
  } else if (existingRule !== null && newRule !== null) {
    // updating existing rule
    const existingIndex = config.data.rules.indexOf(existingRule);

    if (existingIndex >= 0) {
      config.data.rules.splice(existingIndex, 1, newRule);
    }
  }

  return config.data.rules;
};

export const isScheduleUnique = (newSchedule: Schedule, existingSchedule: Schedule | null): boolean => {
  if (existingSchedule !== null) {
    const otherSchedules = config.data.schedules.filter((item) => existingSchedule.name !== item.name);
    const matchingSchedules = otherSchedules.filter((item) => newSchedule.name === item.name);
    return matchingSchedules.length == 0;
  } else {
    const matchingSchedules = config.data.schedules.filter((item) => newSchedule.name === item.name);
    return matchingSchedules.length == 0;
  }
};

export const isRuleUnique = (newRule: Rule, existingRule: Rule | null): boolean => {
  if (existingRule !== null) {
    const otherRules = config.data.rules.filter((item) => existingRule.id !== item.id);
    const matchingRules = otherRules.filter((item) => newRule.name === item.name);
    return matchingRules.length == 0;
  } else {
    const matchingRules = config.data.rules.filter((item) => newRule.name === item.name);
    return matchingRules.length == 0;
  }
};

export const publishSchedules = async (): Promise<number> => {
  // return values: 0=success, 2=version error, 3=failure, 4=in available activity
  if (Manager.getInstance().store.getState().flex.worker.activity.available === true) {
    Notifications.showNotification(NotificationIds.PUBLISH_FAILED_ACTIVITY);
    return 4;
  }

  const latestData: ScheduleManagerConfig = {
    data: {
      rules: config.data.rules.filter((rule) => rule.isDeleted !== true),
      schedules: config.data.schedules.filter((schedule) => schedule.isDeleted !== true),
    },
    version: config.version,
    versionIsDeployed: config.versionIsDeployed,
  };
  latestData.data.rules.map((rule) => {
    rule.isPublished && delete rule.isPublished;
    rule.isDeleted && delete rule.isDeleted;
  });
  latestData.data.schedules.map((schedule) => {
    schedule.isPublished && delete schedule.isPublished;
    schedule.isDeleted && delete schedule.isDeleted;
  });

  const updateResponse = await ScheduleManagerService.update(latestData);

  if (!updateResponse.success) {
    console.log('Schedule update failed', updateResponse);
    ErrorManager.createAndProcessError('Schedule update failed', {
      type: FlexPluginErrorType.action,
      context: 'Plugin.ScheduleManager.publishSchedules',
    });

    if (updateResponse.buildSid == 'versionError') {
      Notifications.showNotification(NotificationIds.PUBLISH_FAILED_OTHER_UPDATE);
      return 2;
    }

    Notifications.showNotification(NotificationIds.PUBLISH_FAILED);
    return 3;
  }

  // the build will take several seconds. use delay and check in a loop.
  await delay(2000);
  let updateStatus = await ScheduleManagerService.updateStatus(updateResponse.buildSid);

  while (updateStatus.buildStatus !== 'completed') {
    if (updateStatus.buildStatus === 'failed' || updateStatus.buildStatus === 'error') {
      // oh no
      console.log('Schedule update build failed', updateStatus);
      ErrorManager.createAndProcessError('Schedule update build failed', {
        type: FlexPluginErrorType.action,
        context: 'Plugin.ScheduleManager.publishSchedules',
      });
      Notifications.showNotification(NotificationIds.PUBLISH_FAILED);
      return 3;
    }

    await delay(2000);
    updateStatus = await ScheduleManagerService.updateStatus(updateResponse.buildSid);
  }

  const publishResponse = await ScheduleManagerService.publish(updateResponse.buildSid);

  if (!publishResponse.success) {
    console.log('Schedule publish failed', publishResponse);
    ErrorManager.createAndProcessError('Schedule publish failed', {
      type: FlexPluginErrorType.action,
      context: 'Plugin.ScheduleManager.publishSchedules',
    });
    Notifications.showNotification(NotificationIds.PUBLISH_FAILED);
    return 3;
  }

  Notifications.showNotification(NotificationIds.PUBLISH_SUCCESS);
  return 0;
};

const checkDate = (rule: Rule, now: Date) => {
  // first, check the date component of the rule.

  // if bounds are provided, check current date against them
  if (rule.startDate) {
    const startDate = new Date(Date.parse(rule.startDate));

    if (now.valueOf() < startDate.valueOf()) {
      // before start date; this rule is not a match
      console.log(`Rule ${rule.name} doesn't match: before start date`, startDate);
      return false;
    }
  }

  if (rule.endDate) {
    const endDate = new Date(Date.parse(rule.endDate));

    if (now.valueOf() > endDate.valueOf()) {
      // after end date; this rule is not a match
      console.log(`Rule ${rule.name} doesn't match: after end date`, endDate);
      return false;
    }
  }

  if (rule.dateRRule) {
    // extract rule from settings
    const ruleOptions = RRule.parseString(rule.dateRRule);

    // since the rule concerns only dates, and we are only comparing to today, we only need one occurrence
    // by default, past occurrences are not returned
    ruleOptions.count = 1;
    ruleOptions.dtstart = now;

    const rrule = new RRule(ruleOptions);
    let matchFound = false;

    rrule.all().forEach((occurrence) => {
      if (now.valueOf() == occurrence.valueOf()) {
        console.log(`Rule ${rule.name} occurrence matched`, occurrence);
        matchFound = true;
      }
    });

    if (!matchFound) {
      console.log(`Rule ${rule.name} doesn't match: no recurrence match`, ruleOptions);
      return false;
    }
  }

  return true;
};

const checkTime = (rule: Rule, nowTz: DateTime) => {
  if (rule.startTime) {
    const startTime = DateTime.fromISO(rule.startTime);

    if (nowTz.hour < startTime.hour) {
      console.log(`Rule ${rule.name} doesn't match: before start hour`, startTime.hour);
      return false;
    }

    if (nowTz.hour == startTime.hour && nowTz.minute < startTime.minute) {
      console.log(`Rule ${rule.name} doesn't match: before start minute`, startTime.minute);
      return false;
    }
  }

  if (rule.endTime) {
    const endTime = DateTime.fromISO(rule.endTime);

    if (nowTz.hour > endTime.hour) {
      console.log(`Rule ${rule.name} doesn't match: after end hour`, endTime.hour);
      return false;
    }

    if (nowTz.hour == endTime.hour && nowTz.minute >= endTime.minute) {
      console.log(`Rule ${rule.name} doesn't match: after end minute`, endTime.minute);
      return false;
    }
  }

  return true;
};

export const evaluateSchedule = (scheduleData: Schedule, rulesData: Rule[]) => {
  const matchingRules: Rule[] = [];
  const returnData = {
    isOpen: false,
    closedReason: 'closed',
  };

  // if the schedule is manually closed, we can return immediately
  if (scheduleData.manualClose) {
    returnData.isOpen = false;
    returnData.closedReason = 'manual';
    return returnData;
  }

  // get the current date/time in the schedule's timezone
  const nowTz: DateTime = DateTime.utc().setZone(scheduleData.timeZone);

  const now = new Date(Date.UTC(nowTz.year, nowTz.month - 1, nowTz.day));

  // evaluate each schedule rule to see if it is currently matching
  scheduleData.rules.forEach((ruleId: string) => {
    const rule = rulesData.find((rule: Rule) => rule.id == ruleId);
    if (rule && checkDate(rule, now) && checkTime(rule, nowTz)) {
      matchingRules.push(rule);
    }
  });

  const closedMatches = matchingRules.filter((rule) => !rule.isOpen);

  // if any non-open schedule(s) match, we are closed
  if (closedMatches.length > 0) {
    returnData.isOpen = false;
    returnData.closedReason = closedMatches[0].closedReason;
  } else {
    const openMatches = matchingRules.filter((rule) => rule.isOpen);

    // if only open schedule(s) match, we are open
    if (openMatches.length > 0) {
      returnData.isOpen = true;
      returnData.closedReason = '';
    }
  }

  // if nothing matches, the default returnData is used

  return returnData;
  // }
};
