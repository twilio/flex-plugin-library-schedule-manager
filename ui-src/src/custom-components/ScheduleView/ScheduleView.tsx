/* eslint-disable no-empty-pattern */
import React, { useEffect, useState } from 'react';
import { Manager, Notifications, Tab, Tabs, withTheme } from '@twilio/flex-ui';
import { Button } from '@twilio-paste/core/button';
import { Heading } from '@twilio-paste/core/heading';
import { Modal, ModalBody } from '@twilio-paste/core/modal';
import { Spinner } from '@twilio-paste/core/spinner';
import { Stack } from '@twilio-paste/core/stack';
import { Text } from '@twilio-paste/core/text';
import AlertBox from '../common/AlertBox';
import { analytics, Event } from '../../utils/Analytics';

import {
  PublishModalContent,
  ScheduleViewWrapper,
  ScheduleViewHeader,
  PublishActionContainer,
} from './ScheduleViewStyles';

import RuleDataTable from '../RuleDataTable/RuleDataTable';
import ScheduleDataTable from '../ScheduleDataTable/ScheduleDataTable';
import { Rule, Schedule } from '../../types/schedule-manager';
import { loadScheduleData, publishSchedules } from '../../utils/schedule-manager';
import { NotificationIds } from '../../flex-hooks/notifications/ScheduleManager';
import ScheduleManagerStrings, { StringTemplates } from '../../flex-hooks/strings/ScheduleManager';
import { Box, Callout, CalloutHeading, CalloutText, Disclosure, DisclosureContent, DisclosureHeading, ListItem, Toaster, UnorderedList, useToaster } from '@twilio-paste/core';

const ScheduleView = ({}) => {
  const [isLoading, setIsLoading] = useState(true);
  const [isChangesPresent, setChangesPresent] = useState(false);
  const toaster = useToaster();
  const [rules, setRules] = useState([] as Rule[]);
  const [schedules, setSchedules] = useState([] as Schedule[]);
  const [updated, setUpdated] = useState(new Date());
  const [isVersionMismatch, setIsVersionMismatch] = useState(false);
  const [selectedTabName, setSelectedTabName] = React.useState<string>('schedules');
  const [loadFailed, setLoadFailed] = useState(false);
  const [publishState, setPublishState] = useState(0); // 0: normal; 1: publish in progress; 2: publish version error; 3: publish failed; 4: in available activity
  const [isDirty, setIsDirty] = useState(false);
  const [cancelSchedule, setCancelSchedule] = useState(false);

  useEffect(() => {
    listSchedules();
    analytics.page(Event.OPHRS_SCHEDULE_LIST);
    return () => {
      if (publishState == 1) {
        Notifications.showNotification(NotificationIds.PUBLISH_ABORTED);
      }
    };
  }, []);

  const listSchedules = async () => {
    setIsLoading(true);

    const scheduleData = await loadScheduleData();

    if (scheduleData === null) {
      setLoadFailed(true);
    } else {
      setLoadFailed(false);
      setRules(scheduleData.data.rules);
      setSchedules(scheduleData.data.schedules);
      setUpdated(new Date());
      setIsVersionMismatch(scheduleData.versionIsDeployed === false);
    }

    setIsLoading(false);
  };

  useEffect(() => {
    const unpublishedSchedules = schedules.filter(
      (schedule) => schedule.isPublished === false || schedule.isDeleted === true,
    );
    const unpublishedRules = rules.filter((rule) => rule.isPublished === false || rule.isDeleted === true);
    setChangesPresent(unpublishedSchedules.length > 0 || unpublishedRules.length > 0);
  }, [schedules, rules]);

  const updateSchedules = (newSchedules: Schedule[]) => {
    setSchedules(newSchedules);
  };

  const updateRules = (newRules: Rule[]) => {
    setRules(newRules);
  };

  const publish = async () => {
    analytics.track(Event.OPHRS_PUBLISH_SCHEDULE);
    setPublishState(1);
    const publishResult = await publishSchedules();
    setPublishState(publishResult);

    switch (publishResult) {
      case 0:
        setIsDirty(false);
        toaster.push({
          message: <Box as="p">{ScheduleManagerStrings[StringTemplates.SUCCESFULLY_PUBLISHED_CHANGES]}</Box>,
          variant: 'success',
          dismissAfter: 4000,
        });
        await listSchedules();
        break;
      case 2:
        toaster.push({
          message: <Box as="p">{ScheduleManagerStrings[StringTemplates.PUBLISH_FAILED_OTHER_UPDATE]}</Box>,
          variant: 'error',
          dismissAfter: 4000,
        });
        break;
      case 3:
        toaster.push({
          message: <Box as="p">{ScheduleManagerStrings[StringTemplates.PUBLISH_FAILED]}</Box>,
          variant: 'error',
          dismissAfter: 4000,
        });
        break;
      case 4:
        toaster.push({
          message: <Box as="p">{ScheduleManagerStrings[StringTemplates.PUBLISH_FAILED_ACTIVITY]}</Box>,
          variant: 'error',
          dismissAfter: 4000,
        });
        break;

      default:
        break;
    }
  };

  const handleCancelConfirmSubmit = (): void => {
    analytics.track(Event.OPHRS_RESET_SCHEDULE);
    listSchedules();
    setCancelSchedule(false);
  };

  return (
    <ScheduleViewWrapper isLight={Manager.getInstance().store.getState().flex.config?.theme?.isLight}>
      <ScheduleViewHeader>
        <Heading as="h3" variant="heading30" marginBottom="space0" data-testid="schedule-manager-title">
          {ScheduleManagerStrings[StringTemplates.SCHEDULE_MANAGER_TITLE]}
        </Heading>
        <Box marginTop={'space30'} className="schedule-manager-info-text">
          <Disclosure>
            <DisclosureHeading as="h2" variant="heading50">
              {ScheduleManagerStrings[StringTemplates.INFO_SECTION_HEADER]}
            </DisclosureHeading>
            <DisclosureContent>
              <UnorderedList>
                <ListItem>{ScheduleManagerStrings[StringTemplates.INFO_SECTION_DESC_TEXT1]}</ListItem>
                <ListItem>{ScheduleManagerStrings[StringTemplates.INFO_SECTION_DESC_TEXT2]}</ListItem>
                <ListItem>{ScheduleManagerStrings[StringTemplates.INFO_SECTION_DESC_TEXT3]}</ListItem>
              </UnorderedList>
            </DisclosureContent>
          </Disclosure>
        </Box>

        {isChangesPresent && (
          <Callout variant="warning" marginY={'space40'}>
            <CalloutHeading as="h2">{ScheduleManagerStrings[StringTemplates.UNPUBLISHED_CHANGES]}</CalloutHeading>
            <CalloutText>
              {ScheduleManagerStrings[StringTemplates.UNPUBLISHED_CHANGES_TEXT]}
              <strong>{ScheduleManagerStrings[StringTemplates.PUBLISH_BUTTON]}</strong>
              {ScheduleManagerStrings[StringTemplates.UNPUBLISHED_CHANGES_TEXT2]}
            </CalloutText>
          </Callout>
        )}
      </ScheduleViewHeader>
      <Tabs
        key="schedule-manager-tabs"
        selectedTabName={selectedTabName}
        onTabSelected={(tabName) => setSelectedTabName(tabName)}
      >
        <Tab label={ScheduleManagerStrings[StringTemplates.TAB_SCHEDULES]} key="schedules" uniqueName="schedules">
          <ScheduleDataTable
            isLoading={isLoading}
            rules={rules}
            schedules={schedules}
            updateSchedules={updateSchedules}
            updated={updated}
            toaster={toaster}
          />
        </Tab>
        <Tab label={ScheduleManagerStrings[StringTemplates.TAB_RULES]} key="rules" uniqueName="rules">
          <RuleDataTable
            isLoading={isLoading}
            rules={rules}
            schedules={schedules}
            updateRules={updateRules}
            toaster={toaster}
          />
        </Tab>
      </Tabs>
      <PublishActionContainer>
        <Stack orientation="horizontal" spacing="space30">
          <Button variant="primary" onClick={publish} disabled={!isChangesPresent}>
            {ScheduleManagerStrings[StringTemplates.PUBLISH_BUTTON]}
          </Button>
          <Button variant="secondary" onClick={() => setCancelSchedule(true)} disabled={!isChangesPresent}>
            {ScheduleManagerStrings[StringTemplates.CANCEL_BUTTON]}
          </Button>
        </Stack>
      </PublishActionContainer>
      <Modal isOpen={publishState === 1} onDismiss={() => {}} size="default" ariaLabelledby="">
        <ModalBody>
          <PublishModalContent>
            <Stack orientation="horizontal" spacing="space60">
              <Spinner decorative={true} size="sizeIcon100" title="Please wait..." />
              <Stack orientation="vertical" spacing="space20">
                <Heading as="h3" variant="heading30" marginBottom="space0">
                  {ScheduleManagerStrings[StringTemplates.PUBLISH_DIALOG_TITLE]}
                </Heading>
                <Text as="p">{ScheduleManagerStrings[StringTemplates.PUBLISH_DIALOG_TEXT]}</Text>
              </Stack>
            </Stack>
          </PublishModalContent>
        </ModalBody>
      </Modal>
      <Modal isOpen={loadFailed} onDismiss={() => {}} size="default" ariaLabelledby="">
        <ModalBody>
          <PublishModalContent>
            <Stack orientation="vertical" spacing="space20">
              <Heading as="h3" variant="heading30" marginBottom="space0">
                {ScheduleManagerStrings[StringTemplates.LOAD_FAILED_TITLE]}
              </Heading>
              <Text as="p">{ScheduleManagerStrings[StringTemplates.LOAD_FAILED_TEXT]}</Text>
            </Stack>
          </PublishModalContent>
        </ModalBody>
      </Modal>
      <AlertBox
        handleClose={() => {
          setCancelSchedule(false);
        }}
        isOpen={!!cancelSchedule}
        handleSubmit={() => handleCancelConfirmSubmit()}
        title={ScheduleManagerStrings[StringTemplates.CANCEL_PUBLISHING_TEXT]}
      >
        <Box as="div" margin={'space10'}>
          <Box as="span">{ScheduleManagerStrings[StringTemplates.CANCEL_PUBLISHING_TEXT2]}</Box>
          <Box as="strong" fontWeight={'fontWeightSemibold'} marginLeft={'space10'}>
            {ScheduleManagerStrings[StringTemplates.CANCEL_PUBLISHING_TEXT3]}
          </Box>
          <Box as="span" marginLeft={'space10'}>
            {ScheduleManagerStrings[StringTemplates.CANCEL_PUBLISHING_TEXT4]}
          </Box>
        </Box>
      </AlertBox>
      <Toaster {...toaster} />
    </ScheduleViewWrapper>
  );
};

export default ScheduleView;
