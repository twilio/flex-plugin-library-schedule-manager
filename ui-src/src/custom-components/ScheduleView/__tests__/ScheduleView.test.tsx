import React from 'react';
import { render, waitFor } from '@testing-library/react';
import ScheduleView from '../ScheduleView';
import * as utils from '../../../utils/schedule-manager';

jest.mock('@twilio-paste/core', () => {
  return {
    ...jest.requireActual('@twilio-paste/core'),
    Toaster: () => <div data-testid="toaster"></div>,
  };
});

jest.mock('../../common/Sortable/SortableList', () => {
  return {
    SortableList: () => <div data-testid="sortable-list"></div>,
  };
});
describe('ScheduleView component', () => {
  test('renders correct data loaded from serverless', async () => {
    const dataPromise: Promise<any> = new Promise((resolve) => {
      resolve({
        data: {
          rules: [],
          schedules: [
            {
              name: 'Example',
              manualClose: false,
              timeZone: 'America/Chicago',
              rules: [],
              status: {
                isOpen: false,
                closedReason: 'closed',
              },
            },
            {
              name: 'Closed on Diwali',
              manualClose: false,
              timeZone: 'Asia/Kolkata',
              rules: [],
              status: {
                isOpen: false,
                closedReason: 'closed',
              },
            },
          ],
        },
        version: 'ZNd6b4b840cc2e20664cd9c7895487261d',
        versionIsDeployed: true,
      });
    });
    const listMock = jest.spyOn(utils, 'loadScheduleData').mockReturnValue(dataPromise);
    const scheduleView = render(<ScheduleView />);
    await waitFor(() => {
      expect(scheduleView).toMatchSnapshot();
      expect(scheduleView.getByTestId('schedule-manager-title').textContent).toBe('Operating hours');
    });

    expect(listMock).toHaveBeenCalledTimes(1);
  });
});
