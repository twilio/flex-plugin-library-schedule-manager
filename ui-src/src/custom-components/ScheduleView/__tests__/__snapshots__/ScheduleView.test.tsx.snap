// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScheduleView component renders correct data loaded from serverless 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="css-zylmng"
      >
        <div
          class="css-b3ml4e"
        >
          <h3
            class="css-n8ur01"
            data-paste-element="HEADING"
            data-testid="schedule-manager-title"
          >
            Schedule manager
          </h3>
          <div
            class="schedule-manager-info-text css-1xorfws"
            data-paste-element="BOX"
          >
            <div
              data-testid="disclosure"
            >
              <div
                data-testid="disclosure-heading"
              >
                Setting up open and closed hours
              </div>
              <div
                data-testid="disclosure-content"
              >
                <ul
                  class="css-17dge34"
                  data-paste-element="UNORDERED_LIST"
                >
                  <li
                    class="css-1t0yrln"
                    data-paste-element="LIST_ITEM"
                  >
                    Create a schedule for each team of agents that operate in the same hours.
                  </li>
                  <li
                    class="css-1t0yrln"
                    data-paste-element="LIST_ITEM"
                  >
                    A schedule is a collection of rules. A rule can define when the contact center is open, like: Open Monday - Friday 8am - 5pm every week. Or closed, like: Closed December 25th every year.
                  </li>
                  <li
                    class="css-1t0yrln"
                    data-paste-element="LIST_ITEM"
                  >
                    Rule are processed in the order in which they are listed in the schedule.
                  </li>
                </ul>
              </div>
            </div>
          </div>
        </div>
        <div
          class="css-12nie3k"
        >
          <div
            class="css-tfro19"
            data-paste-element="STACK"
          >
            <div
              class="css-ci45q"
              data-paste-element="STACK_CHILD"
            >
              <button
                aria-busy="false"
                class="css-eeqvux"
                data-paste-element="BUTTON"
                disabled=""
                type="button"
              >
                <span
                  class="css-z6jp9v"
                  data-paste-element="BOX"
                >
                  Publish changes
                </span>
              </button>
            </div>
            <div
              class="css-roynbj"
              data-paste-element="STACK_CHILD"
            >
              <button
                aria-busy="false"
                class="css-jvdf5z"
                data-paste-element="BUTTON"
                disabled=""
                type="button"
              >
                <span
                  class="css-z6jp9v"
                  data-paste-element="BOX"
                >
                  Cancel
                </span>
              </button>
            </div>
          </div>
        </div>
        <div
          data-testid="toaster"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="css-zylmng"
    >
      <div
        class="css-b3ml4e"
      >
        <h3
          class="css-n8ur01"
          data-paste-element="HEADING"
          data-testid="schedule-manager-title"
        >
          Schedule manager
        </h3>
        <div
          class="schedule-manager-info-text css-1xorfws"
          data-paste-element="BOX"
        >
          <div
            data-testid="disclosure"
          >
            <div
              data-testid="disclosure-heading"
            >
              Setting up open and closed hours
            </div>
            <div
              data-testid="disclosure-content"
            >
              <ul
                class="css-17dge34"
                data-paste-element="UNORDERED_LIST"
              >
                <li
                  class="css-1t0yrln"
                  data-paste-element="LIST_ITEM"
                >
                  Create a schedule for each team of agents that operate in the same hours.
                </li>
                <li
                  class="css-1t0yrln"
                  data-paste-element="LIST_ITEM"
                >
                  A schedule is a collection of rules. A rule can define when the contact center is open, like: Open Monday - Friday 8am - 5pm every week. Or closed, like: Closed December 25th every year.
                </li>
                <li
                  class="css-1t0yrln"
                  data-paste-element="LIST_ITEM"
                >
                  Rule are processed in the order in which they are listed in the schedule.
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-12nie3k"
      >
        <div
          class="css-tfro19"
          data-paste-element="STACK"
        >
          <div
            class="css-ci45q"
            data-paste-element="STACK_CHILD"
          >
            <button
              aria-busy="false"
              class="css-eeqvux"
              data-paste-element="BUTTON"
              disabled=""
              type="button"
            >
              <span
                class="css-z6jp9v"
                data-paste-element="BOX"
              >
                Publish changes
              </span>
            </button>
          </div>
          <div
            class="css-roynbj"
            data-paste-element="STACK_CHILD"
          >
            <button
              aria-busy="false"
              class="css-jvdf5z"
              data-paste-element="BUTTON"
              disabled=""
              type="button"
            >
              <span
                class="css-z6jp9v"
                data-paste-element="BOX"
              >
                Cancel
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        data-testid="toaster"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
